flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/index/IndexGeneratorFactory.java

```diff
@@ -0,0 +1,294 @@
+						String indexFieldValueStr = indexField.toLocalDate().format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} // DataTypes.TIME
+			else if (indexFieldType == Types.LOCAL_TIME) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						LocalTime indexField = (LocalTime) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} else if (indexFieldType == Types.SQL_TIME) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						Time indexField = (Time) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.toLocalTime().format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} else {
+				throw new TableException(String.format("Unsupported type '%s' found in Elasticsearch dynamic index field, " +
+					"time-related pattern only support types are: %s", indexFieldType, indexHelper.getSupportedDateTypes()));
+			}
+		}
+		// general dynamic index pattern
+		return new DefaultIndexGenerator(index) {
+
+			@Override
+			public String generate(Row row) {
+				Object indexField = row.getField(indexFieldPos);
+				return indexPrefix.concat(indexField == null ? index : indexField.toString()).concat(indexSuffix);
+			}
+		};
+	}
+
+	/**
+	 * Helper class for {@link IndexGeneratorFactory}, this helper can use to validate index field type
+	 * ans parse index format from pattern.
+	 */
+	private static class IndexHelper {
+		private static final Pattern dynamicIndexPattern = Pattern.compile("\\{[^\\{\\}]+\\}?");
+		private static final Pattern dynamicIndexTimeExtractPattern = Pattern.compile(".*\\{.+\\|.*\\}.*");
+		private static final List<TypeInformation> supportedTypes = new ArrayList<>();
+		private static final Map<TypeInformation, String> defaultFormats = new HashMap<>();
+
+		static {
+			//time related types
+			supportedTypes.add(Types.LOCAL_DATE_TIME);
+			supportedTypes.add(Types.SQL_TIMESTAMP);
+			supportedTypes.add(Types.LOCAL_DATE);
+			supportedTypes.add(Types.SQL_DATE);
+			supportedTypes.add(Types.LOCAL_TIME);
+			supportedTypes.add(Types.SQL_TIME);
+			//general types
+			supportedTypes.add(Types.STRING);
+			supportedTypes.add(Types.SHORT);
+			supportedTypes.add(Types.INT);
+			supportedTypes.add(Types.LONG);
+		}
+
+		static {
+			defaultFormats.put(Types.LOCAL_DATE_TIME, "yyyy_MM_dd_HH_mm_ss");
+			defaultFormats.put(Types.SQL_TIMESTAMP, "yyyy_MM_dd_HH_mm_ss");
+			defaultFormats.put(Types.LOCAL_DATE, "yyyy_MM_dd");
+			defaultFormats.put(Types.SQL_DATE, "yyyy_MM_dd");
+			defaultFormats.put(Types.LOCAL_TIME, "HH_mm_ss");
+			defaultFormats.put(Types.SQL_TIME, "HH_mm_ss");
```

Comment:
Why underscore?

Reply:
hyphen may lead  truncation issue and elasitcsearch recommends underscore in index name   and field name.